### ISSUE TITLE
Understanding SCRIMP++ Use Cases

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#112]
I'd like to start thinking about adding SCRIMP++ (see #113). However, I'd like to better understand the use cases before implementing something that people won't actually use. Here are some basic questions to get the conversation started:

- [ ] What computer hardware do you have access to (i.e., number of CPUs, number of GPUs, amount of RAM, etc)?
- [ ] What is the minimum, maximum, and average lengths of your time series data?
- [ ] STUMP/GPU-STUMP are pretty fast, why do your use case need SCRIMP++?
- [ ] What would be your expected workflow if you had access to a SCRIMP++ function?  I'm interested in thinking about what the API would look like (i.e., would STUMPY return a Python generator that computes a percentage of the matrix profile that the user specifies)?
- [ ] Is there a future date in which you would no longer need SCRIMP++ (because it's too late)?

**Comment by seanlaw** [comment]
@mexxexx I am super interested in your need for SCRIMP++ and it looks like you already have some familiarity with the matrixprofile-ts library (which, I believe already has SCRIMP++??). Is there a reason why their library is insufficient for your needs? What is the differentiating factor here? 

Our next step is to implement the additional optimizations that is GPU-STOMPopt, which, I believe, is a precursor to SCRIMP++.

If you have the time, I'd love to learn more. I've posed some starting questions above if you have some time to help me understand your needs.

**Comment by mexxexx** [comment]
@seanlaw Thank you for considering this! I'll try to quickly describe my use case: 

I have a (physical) machine that can be tuned by adjusting the values of some parameters A_1, …, A_n (n is around 10). This tuning affects some secondary parameters B_1, …, B_m that cannot be tuned by hand. The machine continuously produces an output X, and by looking at X in a time window, it is possible to decide if the machine was running stable or unstable.

All this information was logged for the past ~10 years, that is roughly around 25M data points.

The tuning of the machine is really complicated, as it can react very sensibly to parameter adjustments and also their influence is not quite well understood, so specialist interventions are needed to keep the machine running at a reasonable performance. The goal is to train a ML model that can support these interventions and generate some insights into how the parameters are related to the stability. For example we would be interested in something like ‘If you raise A_1, you need to lower A_2 in order for the machine to remain stable’ or ‘raising A_1 will increase B_1 in a few hours’.

Up until now we ignored the time component and only ran some clustering to find out which settings were used when the machine was running stable and which were used when it was running unstable. Sadly, the used settings were are greatly (it could have been stable with A_1=100 and A_1=300) and a usually a single setting could lead to a stable as well as an unstable machine, so the time information is crucial.

Based on some research I came to the conclusion that the Matrix Profile could be very helpful in solving this task. I'll try to answer your questions.

* Length of data: I have data from the last ten years out of which ~70% is relevant. Ideally I could sample it at 1.2 second intervals, so that would be around 200M data points. Probably it would suffice to sample at larger time intervals, maybe even a minute, so 6M points. I am very new to the Matrix Profile, so I'm still experimenting on a way smaller subset of the data (max. one year of data at the moment).
* As I sad, I'm experimenting a lot, so I want to have Matrix Profiles ready in very short amounts of time (under a minute), but I don't care about 100% accuracy, because its more of a prototyping and proof of concept. That is why I want to use the Scrimp++ algorithm, as it offers really good anytime properties.
* Also, because of that I develop on a local machine (intel i5, 8GB ram, no(?) GPU), so I do not want to leverage hardware for this. In theory I have access to large CPU clusters (don't know about GPU), but I would like to wait with this if possible.
* I'm at the organization until the end of August and working alone on this project, so I basically need Scrimp right now or in the near future. Depending on your plans, I could help working on an implementation, because I don't think it is too difficult. Sadly I don't really have the capacities for unit testing and large design concepts.
* Probably the way the matrix profile handles scrimp is the most intuitive one: Passing a maximum amount of time you are willing to wait (or None for the exact result). Then Stumpy could also return a completeness percentage.

I have to say, that I'm really a fan of how you support stumpy, so I would like to switch to it. I don't at the moment, because I need the anytime properties that stomp and stump are sadly missing. What I also find really interesting about your library is the multi dimensional version, because this is something I will need to use if my preliminary tests with the matrix profile look promising. I think once Scrimp++ is implemented, it should be really easy too implement a multidimensional version of it as well, using the same approach as for mStomp.

**Comment by seanlaw** [comment]
@mexxexx Thank you for all of the info and for the kind words! I'm sure I'm preaching to the choir when I say this so don't take this as being directed at you but I want make sure the expectations are set appropriately. 😄 While I can't promise anything regarding fast development (STUMPY isn't my day job), the feedback you've provided is certainly helpful and helps us prioritize what we might pursue next. You'll have to be patient as I'm only one person and, as you've already pointed, am very particular about rigorously testing everything so that users can feel confident that the software is doing what is intended and is performant without sacrificing too much on code readability. So, it may take us months before we can get SCRIMP++ into the queue since other things like documentation, tutorials, and illustrative examples have been slightly neglected.

However, we certainly welcome any help (including clarification of the paper's approach)! Frankly, if you (or others reading this) are willing to help out with adding new [tutorials](https://www.cs.ucr.edu/~eamonn/SCRIMP_ICDM_camera_ready_updated.pdf) or [published examples](https://github.com/TDAmeritrade/stumpy/issues/85), then that might free me up to focus on GPU-STOMPopt and SCRIMP++. But absolutely no pressure though! I'm in frequent contact with the original authors and our goal is to faithfully reproduce their algorithms and examples while ensuring that the code is easy to maintain (hence the 100% test coverage) and it is super straightforward for any user to install the software. In my opinion, anything more than Pip/conda install is already way too much friction for my target audience. 

The great thing about SCRIMP++ is that the output (a full matrix profile), when it reaches 100% completion, should be identical to STOMP so all of those tests are essentially known. I'm not sure that I'd test the intermediate outputs along the way but I've probably not thought this through thoroughly enough and am open to your thoughts too.

Regarding multidimensional data, this is REALLY complex and, generally, I am rather pessimistic about it if you don't have enough hardware resources. The computational cost is huge and, IMHO, doesn't scale well when you have more than 5 dimensions. And the details of the published work is sparse. Pragmatically, I'd just start with one-dimensional SCRIMP++ first (assuming GPU-STOMPopt/STOMPopt is done) and hope that it is enough for most use cases before even considering extending it to multi-dimensional SCRIMP++ but don't let me stop you (or others) from contributing. We welcome all forms of help!

At this stage, as a first attempt/pass, it sounds like we should be targeting a multi-CPU version of SCRIMP++ (that will likely just replace `stumpy.stump()`/`stumpy.stumped()` and where `max_time=None` just gives you the exact matrix profile). This is why I'd like to go from STOMP to STOMPopt to SCRIMP++ as this will allow me to gradually update `stumpy.stump()` to using SCRIMP++ but without losing much if any of the existing performant `stumpy.stump()` code. I hope that makes sense (again, I haven't read through SCRIMP++ in detail yet so forgive me if I'm missing something fundamental here).

**Comment by mexxexx** [comment]
@seanlaw I definitly agree that going step by step is the way to go, and that progress should be made based on use cases. 

What I am working on right now is implementing the fourth paper [Matrix Profile IIII: Using Weakly Labeled Time Series to Predict Outcomes](https://www.cs.ucr.edu/~eamonn/WeaklyLabeledTimeSeries.pdf). The algorithm presented there is implemented neither in stumpy nor in the matrixprofile-ts library. Internally it uses the matrix profile, so I try to implement it in a way that is algorithm independent, e.g. by passing a function pointer. Depending on how promising that looks and how it will be performance wise, I might look into implementing Scrimp myself, because I find the matrixprofile-ts code hard to follow.

Like I said, the matrixprofile-ts library does not offer support for the multidimensional algorithms, and while it has not the highest priority right now, it is something I would like to look into.

**Comment by seanlaw** [comment]
@mexxexx Re: SDTS, I think that's great. Perhaps, once you've pieced it together, you would consider contributing a Tutorial that reproduces one of the figures in the original paper. At this point in time, I'd need to be convinced that something like SDTS should necessarily exist within the STUMPY library as a function. Having said that, an SDTS tutorial would be a great example (to demonstrate to others) of how a matrix profile can be used. Let me know what your thoughts are and if you'd like me to take a look at anything. 

As you may have noticed, I've taken a lot of care to ensure that STUMPY faithfully reproduces the pseudocode in the matrix profile papers. From using the same notation to noting the specific tables/equations within their papers, I think it helps to orient the contributor/developer accordingly. I try to be pretty mindful in combining the best of what Python/PyData has to offer (like using generators in FLOSS) and porting things over from Matlab. Anybody who knows Python and has read the original papers should be able to get up and running fairly quickly (I hope). I try not to take too many liberties in changing their code unless the performance is really bad or, in some rare cases, their original papers are misprinted. I certainly value clean code where possible and it's benefited me so far as it relates to maintenance, making small changes (as in the constant subsequence case), and scalability. Of course, I'm not a trained software developer and so I welcome any and all feedback/criticism as it relates to the coding and usability. I've purposely kept everything as clean isolated functions instead of using OOP and I've been debating over whether it is worth creating a [STUMPY class object](https://github.com/TDAmeritrade/stumpy/issues/82). I think the answers "No". At least for now. What do you think?

As for SCRIMP++, I took a quick look last night at the paper. I'm thinking that maybe for now, SCRIMP++ could sit separately from STUMP (instead of replacing it). I get the feeling that most people probably wouldn't run SCRIMP++ to 100% completion (maybe, say 10% max??). And then, if they know that they'd want the full matrix profile then they'd just run the parallel/distributed STUMP/STUMPED/GPU-STUMP. Do you have any performance expectations for SCRIMP++?

**Comment by mexxexx** [comment]
@seanlaw Considering the tutorial, I'm not sure if/when I'll find time for it. But I think that I will be able to publish part of the data I'm using and if I have psoitive results I could then write a small tutorial using SDTS.

I think it is good to stick to the pseudocode, but I have seen various things they omit there, even when they talk about it in the rest of the paper. One example is again SDTS which they do for a fixed subsequence length m, but then explain how it would work for a range. So I think while it makes sense to follow their ideas, I think that some changes and enhancement can definitely be made if they serve a purpose. I think SDTS would have a place in stumpy, because as far as I understood it is like floss an algorithm that builds upon the matrixprofile and I dont think it makes sense to do this in a separate library. Maybe two submodules like stumpy.base and stumpy.algorithms would make sense.

I agree that having a STUMPY class object seems overkill, but maybe some OOP might be helpful. For example I like how scikit-learn has solved the issue. They have a lot of different algorithms that serve the same purpose and that have the same syntax.

```
instance = Algorithm(configuration_parameters)
result = instance.run(data)
```

Such a design would also fit stumpy, where the different algorithms would be stamp, stump, scrimp, ... The good thing about that would be the super easy exchangability. One could just build an algorithm on top of the matrix profile and the user could decide which algorithm and what specific settings he wants to use.

I agree with your last statement that scrimp would be more of a development algorithm if the processing power is available. It still would be interesting to see, how Scrimp performs against the CPU/GPU versions of Stomp. I'm not sure if they did something like this and I also don't know about the possibilities to parallelize scrimp.

**Comment by seanlaw** [comment]
@mexxexx 

> Considering the tutorial, I'm not sure if/when I'll find time for it. But I think that I will be able to publish part of the data I'm using and if I have positive results I could then write a small tutorial using SDTS.

I'm sure you are busy so no pressure at all. For tutorials, we like to start with a Jupyter notebook in the `docs` directory that represents a "work in progress". It doesn't have to have a ton of explanations and text but, as a start, we want to see functioning code that reproduces an example in the original paper. It is enough to reproduce a plot in the paper. There is already a work-in-progress notebook for [multi-dimensional STOMP](https://github.com/TDAmeritrade/stumpy/blob/master/docs/Tutorial_Multidimensional_Motif_Discovery.ipynb). We can take our time polishing it, embellishing it, and adding other examples but the key thing to start is code and an example from the paper. Take a look at the mSTOMP notebook and let me know if this makes sense. 

> I think it is good to stick to the pseudocode, but I have seen various things they omit there, even when they talk about it in the rest of the paper. One example is again SDTS which they do for a fixed subsequence length m, but then explain how it would work for a range. So I think while it makes sense to follow their ideas, I think that some changes and enhancement can definitely be made if they serve a purpose. I think SDTS would have a place in stumpy, because as far as I understood it is like floss an algorithm that builds upon the matrixprofile and I dont think it makes sense to do this in a separate library. Maybe two submodules like stumpy.base and stumpy.algorithms would make sense.

If it's like `floss` then I'd agree. I am completely open to having `stumpy.sdts()` function and throwing everything in an `sdts.py` module. I'm excited for your potential progress!

> I agree that having a STUMPY class object seems overkill, but maybe some OOP might be helpful. For example I like how scikit-learn has solved the issue. They have a lot of different algorithms that serve the same purpose and that have the same syntax.

Agreed. Many API decisions that I've made in STUMPY are inspired/motivated by scikit-learn's simple and user friendly interface. I'd like to revisit the OOP component in the future, perhaps, once we start having too many functions and have enough critical mass in users who'd advocate for this addition.

> I agree with your last statement that scrimp would be more of a development algorithm if the processing power is available. It still would be interesting to see, how Scrimp performs against the CPU/GPU versions of Stomp. I'm not sure if they did something like this and I also don't know about the possibilities to parallelize scrimp.

As with all of the past matrix profile algorithms that were published, the pseudocode always contains some challenging notation (consistent with Matlab) when trying to port them over to programming languages with zero-based indexing. I'm close to figuring it out for SCRIMP (no preSCRIMP yet) so you might see a naive implementation pushed shortly. It will take a bit more effort to:

1. add parallel SCRIMP (via Numba) - but this shouldn't be too much work as this is super parallel-friendly
2. add a timer to stop the JIT compiled SCRIMP function and return the results - this will require some thinking and SO searching but I have some leads

**Comment by seanlaw** [comment]
@mexxexx I was able to add SCRIMP (preSCRIMP to come) in the latest version of the code and it has some tests in place. Since SCRIMP is being JIT compiled for parallel computation, it's not obvious how one would force it to stop running since we'd need Python to tell the JIT compiled code to stop and then return back the intermediate results. There were some ugly threading approaches that I came across but it just wasn't readable. Instead, I opted for a "percentage" argument that approximately stops after a minimum percentage of distance calculations have been computed. Would you mind taking it for a test drive and providing any feedback that you may have?

**Comment by zpzim** [comment]
It looks like you have found a use case for something faster :)

For the anytime property, you could also try implementing something like SCAMP's tiling. Where pieces of the distance matrix are computed at once, so you can stop anytime between a tile.

The key here is to execute in small batches, but large enough to where you don't pay a performance overhead for not giving enough work to the machine.

**Comment by mexxexx** [comment]
@seanlaw Thank you very much. I will look at the implementation and test it out in the coming days!

**Comment by seanlaw** [comment]
@mexxexx Is this still high on your priority? I've been chipping away at SCRIMP and plan to work on adding NaN/inf support (mostly just adding unit tests), then converting `stumpy.scrimp` into a Python generator so that one can incrementally update the matrix profile (so know that the interface will be changing soon), and then I'll work on the preSCRIMP algorithm.  

However, if this is no longer a priority then I will focus on docs/tutorials first and slowly chip away at it.

**Comment by mexxexx** [comment]
Hi @seanlaw, sadly I don't see how I can use the matrix profile in my work, so this doesn't have high priority anymore. Thank you very much for the work you put in! However, if you feel like you need support please feel free to ask, I'd like to help add some new functionality.

**Comment by seanlaw** [comment]
> sadly I don't see how I can use the matrix profile in my work, so this doesn't have high priority anymore. Thank you very much for the work you put in! 

That's too bad but I completely understand. Let's continue to chat in this [separate issue](https://github.com/TDAmeritrade/stumpy/issues/163)
